### PR TITLE
Feat: do not create Prow jobs for Gerrit patch set with no code change

### PR DESCRIPTION
**Pull request by timwangmusic**
This change attempts to address the issue `Not trigger jobs if only description changed`. For a Gerrit patch set with revision info, we can check the `Kind` field to determine if the patch set contains code change over its parent patch set. If not, we can pass processing the change.
/cc @mpherman2 @listx @cjwagner @chaodaiG

### CODE REVIEW

File: prow/gerrit/adapter/adapter.go [hunk 1]
```diff
@@ -427,6 +427,19 @@ func (c *Controller) processChange(logger logrus.FieldLogger, instance string, c
 		logger.WithField("duration(s)", time.Since(tracker).Seconds()).Debug("Time taken for gerrit change to be processed.")
 	}()
 
+	if change.Revisions != nil {
```
Comment (jmguzik) [inline prow/gerrit/adapter/adapter.go:430]: I think this change would also deserve a test case.

File: prow/gerrit/adapter/adapter.go [hunk 2]
```diff
@@ -427,6 +427,19 @@ func (c *Controller) processChange(logger logrus.FieldLogger, instance string, c
...
+		if rev, ok := change.Revisions[change.CurrentRevision]; ok {
```
Comment (chaodaiG) [inline prow/gerrit/adapter/adapter.go:431]: Prow polls Gerrit periodically, and there could be more than 1 revision in between polls, a more reliable way would be to figure out all the changes since last poll, I think `LastSyncState` is what this is used for

Comment (timwangmusic) [reply prow/gerrit/adapter/adapter.go]: Makes sense, done.

File: prow/gerrit/adapter/adapter.go [hunk 3]
```diff
@@ -419,14 +420,41 @@ func (c *Controller) handleInRepoConfigError(err error, instance string, change
...
 		logger.WithField("duration(s)", time.Since(tracker).Seconds()).Debug("Time taken for gerrit change to be processed.")
 	}()
 
+	if change.Revisions != nil {
+		var revisions []gerrit.RevisionInfo
+		for _, rev := range change.Revisions {
+			revisions = append(revisions, rev)
+		}
+		// sort revisions by create time descending
+		sort.Slice(revisions, func(i, j int) bool {
```
Comment (chaodaiG) [inline prow/gerrit/adapter/adapter.go:437]: I don't think you need to sort, as which one is evaluated earlier doesn't really make a difference. So the next for loop could just be:
```
var hasCodeChange bool
for _, rev := range revisions {
			if rev.Created.Time.After(lastUpdated) && rev.Kind != gerrit.NoCodeChange {
                                 hasCodeChange = true
				break
			}
		}
```

Comment (timwangmusic) [reply prow/gerrit/adapter/adapter.go]: Great simplifications, done.

File: prow/gerrit/adapter/adapter.go [hunk 4]
```diff
@@ -419,14 +419,40 @@ func (c *Controller) handleInRepoConfigError(err error, instance string, change
...
 	defer func() {
 		// tracker will reset in `processChange` function, and this is the time
 		// taken for a gerrit change to be processed after the last reset.
 		logger.WithField("duration(s)", time.Since(tracker).Seconds()).Debug("Time taken for gerrit change to be processed.")
 	}()
 
+	if change.Revisions != nil {
+		var hasTestOrRetestCommand bool
+		for _, message := range currentMessages(change, lastUpdated) {
+			if strings.HasPrefix(message.Message, "/test") || strings.HasPrefix(message.Message, "/retest") {
```
Comment (chaodaiG) [inline prow/gerrit/adapter/adapter.go:433]: this is already checked below, I don't like this but I don't have a better idea. This check is not going to work with multi-line comments, using the regex at https://github.com/kubernetes/test-infra/blob/9aadff258d20bb9fd72055c2a0621afcedb23a54/prow/pjutil/filter.go#L30 would be more reliable

File: prow/gerrit/adapter/adapter.go [hunk 5]
```diff
@@ -419,14 +419,40 @@ func (c *Controller) handleInRepoConfigError(err error, instance string, change
...
+
+		var hasCodeChange bool
+		for _, rev := range change.Revisions {
+			if rev.Created.Time.After(lastUpdated) && rev.Kind != gerrit.NoCodeChange {
+				hasCodeChange = true
+				break
+			}
+		}
+
+		if !hasTestOrRetestCommand && !hasCodeChange {
```
Comment (chaodaiG) [inline prow/gerrit/adapter/adapter.go:447]: I think this will skip postsubmit jobs, might need to move this entire closure to be under `client.New`

File: prow/gerrit/adapter/adapter.go (hunk 1 above)
Comment (timwangmusic) [reply prow/gerrit/adapter/adapter.go:430]: Done.

File: prow/gerrit/adapter/adapter.go (hunk 4 above)
Comment (timwangmusic) [reply prow/gerrit/adapter/adapter.go:433]: Added another regex for commands start with `/test`.

File: prow/gerrit/adapter/adapter.go (hunk 5 above)
Comment (timwangmusic) [reply prow/gerrit/adapter/adapter.go:447]: Moved the whole closure to `client.New`.

File: prow/gerrit/adapter/adapter_test.go [hunk 6]
```diff
@@ -513,6 +513,24 @@ func TestProcessChange(t *testing.T) {
 			instance:     testInstance,
 			wantError:    true,
 		},
+		{
```
Comment (chaodaiG) [inline prow/gerrit/adapter/adapter_test.go:516]: will need more tests based on different comments, including corner cases of comments like
```
- /test something
```

File: prow/gerrit/adapter/adapter.go [hunk 7]
```diff
@@ -585,6 +585,32 @@ func (c *Controller) processChange(logger logrus.FieldLogger, instance string, c
 			logger.WithField("lastUpdate", lastUpdate).Warnf("lastUpdate not found, falling back to now")
 		}
 
+		if change.Revisions != nil {
```
Comment (chaodaiG) [inline prow/gerrit/adapter/adapter.go:588]: the purpose of this PR is to skip the inrepoconfig processing, this check will need to be moved to before line 537 to work as expected

File: prow/gerrit/adapter/adapter.go [hunk 8]
```diff
@@ -585,6 +585,32 @@ func (c *Controller) processChange(logger logrus.FieldLogger, instance string, c
...
+
+			var hasCodeChange bool
+			for _, rev := range change.Revisions {
+				if rev.Created.Time.After(lastUpdate) && rev.Kind != gerrit.NoCodeChange {
+					hasCodeChange = true
+					break
+				}
+			}
+
+			if !hasTestOrRetestCommand && !hasCodeChange {
```
Comment (chaodaiG) [inline prow/gerrit/adapter/adapter.go:605]: I'm not a big fan of defaulting to skip processing, it could fail unexpectedly in the future.

An alternative idea is something like:
```
var shouldSkipProcessing bool
for _, message := range currentMessages(change, lastUpdate) {
  if pjutil.RetestRe.Match([]byte(message.Message)) || pjutil.TestRe.Match([]byte(message.Message)) {
	  shouldSkipProcessing = false
	  break
  }
}

// check code change
...

if !shouldSkipProcessing {
  return nil;
}
```
the idea is that the for loop needs to be exercised so that the test can be skipped

File: prow/gerrit/adapter/adapter.go (hunk 7 above)
Comment (timwangmusic) [reply prow/gerrit/adapter/adapter.go:588]: Done.

File: prow/gerrit/adapter/adapter_test.go (hunk 6 above)
Comment (timwangmusic) [reply prow/gerrit/adapter/adapter_test.go:516]: Done.

File: prow/gerrit/adapter/adapter.go (hunk 8 above)
Comment (chaodaiG) [reply prow/gerrit/adapter/adapter.go]: > since the zero value of `bool` is `false`, we have to set the initial value of `shouldSkipProcessing` to true anyway.

sorry but I disagree. The desired default should be a safe fallback, so even if everything is wrong:
```
var shouldSkipProcessing bool

// Complex logics with shouldSkipProcessing
...

if shouldSkipProcessing {
  // This code block should not be triggered even if the code above in the black box failed miserably
  return nil
}
```

As discussed offline that this PR is best effort to save time spent on inrepoconfig processing, so the bottom line is that if there is any bug in this logic then inrepoconfig should still be processed, and users won't tell the difference. The last thing we want to see is user updated a PR and Prow jobs not triggered because of bug here.

Comment (chaodaiG) [reply prow/gerrit/adapter/adapter.go]: @timwangmusic , please take a look at this comment

Comment (chaodaiG) [reply prow/gerrit/adapter/adapter.go]: Had one more offline discussion with Tim, realized that it's also ugly to implement the way that I proposed. It might be easier to move this short closure to a function for easier testing

File: prow/gerrit/adapter/adapter.go [hunk 9]
```diff
@@ -692,6 +699,28 @@ func (c *Controller) processChange(logger logrus.FieldLogger, instance string, c
 	return nil
 }
 
+// shouldSkipChangeProcessing returns true if the revisions in the change after
+// last repo update does not contain code change or valid test related comments.
+func shouldSkipChangeProcessing(change client.ChangeInfo, lastUpdate time.Time) bool {
+	if change.Revisions == nil {
```
Comment (chaodaiG) [inline prow/gerrit/adapter/adapter.go:705]: nit: might want to add a comment stating that this should not happen

File: prow/gerrit/adapter/adapter.go [hunk 10]
```diff
@@ -692,6 +699,28 @@ func (c *Controller) processChange(logger logrus.FieldLogger, instance string, c
...
+		return false
+	}
+
+	for _, message := range currentMessages(change, lastUpdate) {
+		if pjutil.RetestRe.Match([]byte(message.Message)) || pjutil.TestRe.Match([]byte(message.Message)) {
```
Comment (chaodaiG) [inline prow/gerrit/adapter/adapter.go:710]: I think you can use `MatchString`

**Review by chaodaiG — APPROVED** [review]
generally lgtm
two nits
/hold

File: prow/gerrit/adapter/adapter.go (hunk 10 above)
Comment (timwangmusic) [reply prow/gerrit/adapter/adapter.go:710]: done

**Review by chaodaiG — APPROVED** [review]
